includes/class-wc-stripe-order.php

```diff
@@ -169,44 +169,76 @@ class WC_Stripe_Order extends WC_Order {
 	 */
 	const PAYMENT_LOCK_EXPIRATION = 5 * MINUTE_IN_SECONDS;
 
+	/**
+	 * The Stripe Order instances.
+	 *
+	 * @var array
+	 */
+	private static $instances = [];
+
 	/**
 	 * Converts an order into WC_Stripe_Order if it is not already.
 	 *
 	 * @param $order WC_Stripe_Order|WC_Order Order object.
+	 * @param bool $cached Whether to use cached instance.
 	 * @return WC_Stripe_Order
 	 */
-	public static function to_instance( $order ) {
-		return $order instanceof WC_Stripe_Order ? $order : new self( $order );
+	public static function to_instance( $order, $cached = true ) {
+		if ( $order instanceof WC_Stripe_Order ) {
+			return $order;
+		}
+
+		$order_id = $order->get_id();
+		if ( ! $cached || ! isset( self::$instances[ $order_id ] ) ) {
+			self::$instances[ $order_id ] = new self( $order );
+		}
+
+		return self::$instances[ $order_id ];
+	}
+
+	/**
+	 * Clears the cache of instances.
+	 *
+	 * @return void
+	 */
+	public static function clear_cache() {
+		self::$instances = [];
 	}
 
 	/**
 	 * Wrapper to create an order using the extension's custom WC_Stripe_Order class.
 	 *
 	 * @param $order_data array Order data.
+	 * @param bool $cache Whether to cache the instance.
 	 * @return bool|WC_Stripe_Order
 	 */
-	public static function create( $order_data = [] ) {
+	public static function create( $order_data = [], $cache = true ) {
 		$order = wc_create_order( $order_data );
 		if ( ! $order ) {
 			return false;
 		}
 
-		return new WC_Stripe_Order( $order );
+		return self::to_instance( $order, ! $cache );
 	}
 
 	/**
 	 * Wrapper to return an order using the extension's custom WC_Stripe_Order class.
 	 *
 	 * @param $order_id int Order ID.
+	 * @param bool $cached Whether to use cached instance.
 	 * @return bool|WC_Stripe_Order
 	 */
-	public static function get_by_id( $order_id ) {
+	public static function get_by_id( $order_id, $cached = true ) {
+		if ( $cached && isset( self::$instances[ $order_id ] ) ) {
+			return self::$instances[ $order_id ];
+		}
+
 		$order = wc_get_order( $order_id );
 		if ( ! $order ) {
 			return false;
 		}
 
-		return new WC_Stripe_Order( $order );
+		return self::to_instance( $order, $cached );
```

Comment:
Even if we don't move forward with this cache system, that's a change we should keep.